File: Samples CSharp/Advanced/CategoryDesigner/TranslationItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using Therefore.API;

namespace Therefore.Samples
{
    public abstract class TranslationItem
    {
        public Dictionary<int, String> Translations
        {
            get;set;
        }

        public virtual String ItemType
        {
            get;
        }

        public abstract void SetTranslatedValues();
    }

    public class CategoryNameTranslationItem : TranslationItem
    {
        private TheCategory _category;

        public CategoryNameTranslationItem(TheCategory cat)
        {
            Translations = new Dictionary<int, string>();
            _category = cat;
            TheIntArray lcids = cat.GetAvailableLCIDs();
            foreach (int lcid in lcids)
                Translations.Add(lcid, cat.GetName(lcid));
        }

        public override String ItemType { get { return "Category Name"; } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                if (Translations[lcid] == null)
                    throw new Exception(String.Format("Invalid category name for language {0}", CultureInfo.GetCultureInfo(lcid).Name));
                _category.SetName(lcid, Translations[lcid]);
            }
        }
    }

    public class CategoryDescriptionTranslationItem : TranslationItem
    {
        private TheCategory _category;

        public CategoryDescriptionTranslationItem(TheCategory cat)
        {
            Translations = new Dictionary<int, string>();
            _category = cat;
            TheIntArray lcids = cat.GetAvailableLCIDs();
            foreach (int lcid in lcids)
                Translations.Add(lcid, cat.GetDescription(lcid));
        }

        public override String ItemType { get { return "Category Description"; } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                String description = "";
                if (Translations[lcid] != null)
                    description = Translations[lcid];
                _category.SetDescription(lcid, description);
            }
        }
    }

    public class CaseDefinitionNameTranslationItem : TranslationItem
    {
        private TheCaseDefinition _casedef;

        public CaseDefinitionNameTranslationItem(TheCaseDefinition casedef)
        {
            Translations = new Dictionary<int, string>();
            _casedef = casedef;
            TheIntArray lcids = casedef.GetAvailableLCIDs();
            foreach (int lcid in lcids)
                Translations.Add(lcid, casedef.GetName(lcid));
        }

        public override String ItemType { get { return "Case Definition Name"; } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                if (Translations[lcid] == null)
                    throw new Exception(String.Format("Invalid case definition name for language {0}", CultureInfo.GetCultureInfo(lcid).Name));
                _casedef.SetName(lcid, Translations[lcid]);
            }
        }
    }

    public class CaseDefinitionDescriptionTranslationItem : TranslationItem
    {
        private TheCaseDefinition _casedef;

        public CaseDefinitionDescriptionTranslationItem(TheCaseDefinition casedef)
        {
            Translations = new Dictionary<int, string>();
            _casedef = casedef;
            TheIntArray lcids = casedef.GetAvailableLCIDs();
            foreach (int lcid in lcids)
                Translations.Add(lcid, casedef.GetDescription(lcid));
        }

        public override String ItemType { get { return "Case Definition Description"; } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                String description = "";
                if (Translations[lcid] != null)
                    description = Translations[lcid];
                _casedef.SetDescription(lcid, description);
            }
        }
    }

    public class CategoryFieldTranslationItem : TranslationItem
    {
        private TheCategoryField _field;

        public CategoryFieldTranslationItem(TheCategoryField field)
        {
            Translations = new Dictionary<int, string>();
            _field = field;
        }
        
        public override String ItemType { get { return _field.FieldType.ToString() + " (" + _field.FieldNo.ToString() + ")"; } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                if (Translations[lcid] == null)
                    throw new Exception(String.Format("Invalid caption for field {0} language {1}", _field.FieldNo, CultureInfo.GetCultureInfo(lcid).Name));
                _field.SetCaption(lcid, Translations[lcid]);
            }
        }
    }

    public class TabInfoTranslationItem : TranslationItem
    {
        private TheTabInfo _tabinfo;
        private int _fieldno;

        public TabInfoTranslationItem(TheTabInfo tabinfo, int fieldno)
        {
            Translations = new Dictionary<int, string>();
            _tabinfo = tabinfo;
            _fieldno = fieldno;
        }

        public override String ItemType { get { return "TabControlField " + _fieldno + " Tab " + _tabinfo.TabNo.ToString(); } }

        public override void SetTranslatedValues()
        {
            foreach (int lcid in Translations.Keys)
            {
                if (Translations[lcid] == null)
                    throw new Exception(String.Format("Invalid caption for tab {0} of field {1} language {2}", _tabinfo.TabNo, _fieldno, CultureInfo.GetCultureInfo(lcid).Name));
                _tabinfo.SetCaption(lcid, Translations[lcid]);
            }
        }
    }
}

File: Samples VB.NET/Advanced/CategoryDesigner/TranslationItem.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Globalization
Imports Therefore.API

Namespace Therefore.Samples
	Public MustInherit Class TranslationItem
		Public Property Translations() As Dictionary(Of Integer, String)

		Public Overridable ReadOnly Property ItemType() As String

		Public MustOverride Sub SetTranslatedValues()
	End Class

	Public Class CategoryNameTranslationItem
		Inherits TranslationItem
		Private _category As TheCategory

		Public Sub New(ByVal cat As TheCategory)
			Translations = New Dictionary(Of Integer, String)()
			_category = cat
			Dim lcids As TheIntArray = cat.GetAvailableLCIDs()
			For Each lcid As Integer In lcids
				Translations.Add(lcid, cat.GetName(lcid))
			Next lcid
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return "Category Name"
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				If Translations(lcid) Is Nothing Then
					Throw New Exception(String.Format("Invalid category name for language {0}", CultureInfo.GetCultureInfo(lcid).Name))
				End If
				_category.SetName(lcid, Translations(lcid))
			Next lcid
		End Sub
	End Class

	Public Class CategoryDescriptionTranslationItem
		Inherits TranslationItem
		Private _category As TheCategory

		Public Sub New(ByVal cat As TheCategory)
			Translations = New Dictionary(Of Integer, String)()
			_category = cat
			Dim lcids As TheIntArray = cat.GetAvailableLCIDs()
			For Each lcid As Integer In lcids
				Translations.Add(lcid, cat.GetDescription(lcid))
			Next lcid
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return "Category Description"
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				Dim description As String = ""
				If Translations(lcid) IsNot Nothing Then
					description = Translations(lcid)
				End If
				_category.SetDescription(lcid, description)
			Next lcid
		End Sub
	End Class

	Public Class CaseDefinitionNameTranslationItem
		Inherits TranslationItem
		Private _casedef As TheCaseDefinition

		Public Sub New(ByVal casedef As TheCaseDefinition)
			Translations = New Dictionary(Of Integer, String)()
			_casedef = casedef
			Dim lcids As TheIntArray = casedef.GetAvailableLCIDs()
			For Each lcid As Integer In lcids
				Translations.Add(lcid, casedef.GetName(lcid))
			Next lcid
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return "Case Definition Name"
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				If Translations(lcid) Is Nothing Then
					Throw New Exception(String.Format("Invalid case definition name for language {0}", CultureInfo.GetCultureInfo(lcid).Name))
				End If
				_casedef.SetName(lcid, Translations(lcid))
			Next lcid
		End Sub
	End Class

	Public Class CaseDefinitionDescriptionTranslationItem
		Inherits TranslationItem
		Private _casedef As TheCaseDefinition

		Public Sub New(ByVal casedef As TheCaseDefinition)
			Translations = New Dictionary(Of Integer, String)()
			_casedef = casedef
			Dim lcids As TheIntArray = casedef.GetAvailableLCIDs()
			For Each lcid As Integer In lcids
				Translations.Add(lcid, casedef.GetDescription(lcid))
			Next lcid
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return "Case Definition Description"
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				Dim description As String = ""
				If Translations(lcid) IsNot Nothing Then
					description = Translations(lcid)
				End If
				_casedef.SetDescription(lcid, description)
			Next lcid
		End Sub
	End Class

	Public Class CategoryFieldTranslationItem
		Inherits TranslationItem
		Private _field As TheCategoryField

		Public Sub New(ByVal field As TheCategoryField)
			Translations = New Dictionary(Of Integer, String)()
			_field = field
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return _field.FieldType.ToString() & " (" & _field.FieldNo.ToString() & ")"
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				If Translations(lcid) Is Nothing Then
					Throw New Exception(String.Format("Invalid caption for field {0} language {1}", _field.FieldNo, CultureInfo.GetCultureInfo(lcid).Name))
				End If
				_field.SetCaption(lcid, Translations(lcid))
			Next lcid
		End Sub
	End Class

	Public Class TabInfoTranslationItem
		Inherits TranslationItem
		Private _tabinfo As TheTabInfo
		Private _fieldno As Integer

		Public Sub New(ByVal tabinfo As TheTabInfo, ByVal fieldno As Integer)
			Translations = New Dictionary(Of Integer, String)()
			_tabinfo = tabinfo
			_fieldno = fieldno
		End Sub

		Public Overrides ReadOnly Property ItemType() As String
			Get
				Return "TabControlField " & _fieldno & " Tab " & _tabinfo.TabNo.ToString()
			End Get
		End Property

		Public Overrides Sub SetTranslatedValues()
			For Each lcid As Integer In Translations.Keys
				If Translations(lcid) Is Nothing Then
					Throw New Exception(String.Format("Invalid caption for tab {0} of field {1} language {2}", _tabinfo.TabNo, _fieldno, CultureInfo.GetCultureInfo(lcid).Name))
				End If
				_tabinfo.SetCaption(lcid, Translations(lcid))
			Next lcid
		End Sub
	End Class
End Namespace
